File: go/nucl-parquet/interp.go
package nuclparquet

import (
	"math"
	"sort"
)

// LogLogInterp performs log-log interpolation on sorted x/y arrays.
//
// Both slices must be the same length and sorted by x (ascending).
// Values are clamped to the first/last table entry when outside range.
// Returns NaN if the slice is empty or energy <= 0.
func LogLogInterp(xs, ys []float64, x float64) float64 {
	n := len(xs)
	if n == 0 || x <= 0 {
		return math.NaN()
	}

	// Clamp to table range.
	if x <= xs[0] {
		return ys[0]
	}
	if x >= xs[n-1] {
		return ys[n-1]
	}

	// Binary search for the bracketing interval.
	idx := sort.SearchFloat64s(xs, x)

	// Exact match.
	if idx < n && xs[idx] == x {
		return ys[idx]
	}

	// x is between xs[idx-1] and xs[idx].
	i := idx - 1
	x0, x1 := xs[i], xs[i+1]
	y0, y1 := ys[i], ys[i+1]

	// If either value is zero or negative, fall back to linear interpolation.
	if y0 <= 0 || y1 <= 0 {
		t := (x - x0) / (x1 - x0)
		return y0 + t*(y1-y0)
	}

	// Log-log interpolation.
	logX := math.Log(x)
	logX0 := math.Log(x0)
	logX1 := math.Log(x1)
	logY0 := math.Log(y0)
	logY1 := math.Log(y1)

	t := (logX - logX0) / (logX1 - logX0)
	return math.Exp(logY0 + t*(logY1-logY0))
}

File: clients/ts/nucl-parquet/src/interp.ts
/**
 * Log-log interpolation on sorted x/y arrays.
 *
 * Both arrays must be the same length and sorted ascending by x.
 * Returns NaN if x <= 0 or arrays are empty.
 * Clamps to boundary values outside the tabulated range.
 * Falls back to linear interpolation when either bracketing value is <= 0.
 */
export function logLogInterp(xs: Float64Array | number[], ys: Float64Array | number[], x: number): number {
  const n = xs.length;
  if (n === 0 || x <= 0) return NaN;

  // Clamp to table range
  if (x <= xs[0]) return ys[0];
  if (x >= xs[n - 1]) return ys[n - 1];

  // Binary search for the bracketing interval
  let lo = 0;
  let hi = n - 1;
  while (lo < hi - 1) {
    const mid = (lo + hi) >>> 1;
    if (xs[mid] <= x) {
      lo = mid;
    } else {
      hi = mid;
    }
  }

  // Exact match check
  if (xs[lo] === x) return ys[lo];
  if (xs[hi] === x) return ys[hi];

  const e0 = xs[lo];
  const e1 = xs[hi];
  const v0 = ys[lo];
  const v1 = ys[hi];

  // Handle zero or negative values (can't take log)
  if (v0 <= 0 || v1 <= 0) {
    const t = (x - e0) / (e1 - e0);
    return v0 + t * (v1 - v0);
  }

  // Log-log interpolation
  const logX = Math.log(x);
  const logE0 = Math.log(e0);
  const logE1 = Math.log(e1);
  const logV0 = Math.log(v0);
  const logV1 = Math.log(v1);

  const t = (logX - logE0) / (logE1 - logE0);
  return Math.exp(logV0 + t * (logV1 - logV0));
}
